accepted/2022/enable-512-vectors.md

```diff
@@ -0,0 +1,518 @@
+    VectorMask<int> mask = v1.GreaterThanByVectorMask(0) & v1.NotEqualsByVectorMask(5);
+    vresult = Vector.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+Note that the `VectorMask` is built directly expressing the condition of the source vector `v1`. This roughly translates to a scalar loop...
+
+```C#
+for (int i = 0; i < source.Length; i++)
+{
+  if (source[i] > 0 && source[i] != 5)
+    result[i] += source[i];
+}
+```
+
+Importantly, this conditional logic is not limited to comparing `Vector` against constants. In the following example, we perform a conditional add if the elements of `v1` are less than `v2`, i.e., `v2`:
+
+```C#
+public Vector256<int> SumTwoVector256(ReadOnlySpan<int> s1, ReadOnlySpan<int> s2)
+{
+  Vector256<int> vresult = Vector<int>.Zero;
+
+  for (int i = 0; i < s1.Length;; i += Vector256<int>.Count)
+  {
+    Vector256<int> v1 = new Vector256<int>(s1.Slice(i));
+    Vector256<int> v2 = new Vector256<int>(s2.Slice(i));
+
+    VectorMask256<int> mask = v1.LessThanByVectorMask(v2);
+    
+    vresult = Vector256.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+which roughly translates to the following scalar loop logic...
+
+```C#
+for (int i = 0; i < s1.Length; i++)
+{
+  if (s1[i] < 0 && s2[i])
+    result[i] += s1[i];
+}
+```
+
+As `VectorMask256` expresses that we have a masked condition for `Vector256`, so to will we have `VectorMask128`, and `VectorMask64`. In addition, when using the variable length `Vector<T>` API, we have `VectorMask<T>`, as seen the in the following example:
+
+
+```C#
+public Vector<int> SumVector(ReadOnlySpan<int> source)
+{
+  Vector<int> vresult = Vector<int>.Zero;
+
+  for (int i = 0; i < source.Length;; i += Vector<int>.Count)
+  {
+    Vector<int> v1 = new Vector<int>(source.Slice(i));
+    VectorMask<int> mask = v1.GreaterThanByVectorMask(0) & v1.NotEqualByVectorMask(5);
+    vresult = Vector.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+Where `VectorMask<T>` expresses that the number of elements the condition applies to is variable-length, and determined by the JIT at runtime (though it must be compatible with `Vector<T>` selected length).
+
+Lastly, we propose to create a `VectorMask` using builtin C# boolean expressions by passing a lambda to a special `MaskExpr` API:
+
+```C#
+public Vector256<int> SumVector256(ReadOnlySpan<int> source)
+{
+  Vector256<int> vresult = Vector256<int>.Zero;
+
+  for (int i = 0; i < source.Length;; i += Vector256<int>.Count)
+  {
+    Vector256<int> v1 = new Vector256<int>(source.Slice(i));
+    VectorMask256<int> mask = v1.MaskExpr(x => x > 0 & x != 5);
+    vresult = Vector256.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+Logically, the lambda passed to `MaskExpr` selects for which elements of `v1` to include in the `VectorMask`, and allows developers to program conditional SIMD logic with familiar boolean condition operations.
+
+### Leading/Trailing Element Processing with `VectorMask<T>`
```

Comment:
I don't disagree with these ideas, but each represents a slight "layer above" the current way `Vector` API is used. I like the idea of allowing the JIT to perform more powerful vectorization from more declarative programing construct --- the slice, sliceMask you propose is a nice idea --- but it seems that this could be a next step if others agree.